tensorflow/python/ops/array_grad.py

```diff
@@ -354,9 +354,15 @@ def _PreventGradientGrad(op, _):
 def _GatherGrad(op, grad):
   """Gradient for Gather op."""
   # params can be large, so colocate the shape calculation with it.
+  #
+  # params can be super-large for sparse model, array_ops.shape raises
```

Comment:
Here is the exception:
```
>>> import tensorflow as tf
>>> a = tf.get_variable("W", [3000000000], dtype=tf.bool)
>>> b = tf.shape(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\tensorflow\copyable\python3.5\lib\site-packages\tensorflow\python\ops\array_ops.py", line 245, in shape
    return shape_internal(input, name, optimize=True, out_type=out_type)
  File "D:\tensorflow\copyable\python3.5\lib\site-packages\tensorflow\python\ops\array_ops.py", line 271, in shape_i
nternal
    return constant(input_shape.as_list(), out_type, name=name)
  File "D:\tensorflow\copyable\python3.5\lib\site-packages\tensorflow\python\framework\constant_op.py", line 102, in
 constant
    tensor_util.make_tensor_proto(value, dtype=dtype, shape=shape, verify_shape=verify_shape))
  File "D:\tensorflow\copyable\python3.5\lib\site-packages\tensorflow\python\framework\tensor_util.py", line 377, in
 make_tensor_proto
    nparray = np.array(values, dtype=np_dt)
OverflowError: Python int too large to convert to C long
```
Windows numpy throws exception when converting large int to C long while linux will truncate it silently.
```
Python 3.4.3 (default, Nov 17 2016, 01:08:31)
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> a = np.array(3000000000, np.int32)
>>> a
array(-1294967296, dtype=int32)
>>>
``` 
Discussed in #9781 , it's hard to change tf.shape default int32 to int64, also can't change make_tensor_proto, it will make hundreds of dtype inference.